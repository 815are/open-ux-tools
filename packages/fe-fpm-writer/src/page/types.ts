import type { CustomElement, InternalCustomElement } from '../common/types';

/**
 * Configuration options for adding a custom page.
 */
export interface CustomPage extends CustomElement {
    /**
     * Name of the entity used for the custom page.
     */
    entity: string;

    /**
     * Optional incoming navigation configuration. If provided then a navigation configuration is added to an existing page navigating to this custom page.
     */
    navigation?: {
        sourcePage: string;
        sourceEntity: string;
        navEntity: string;
    };

    /**
     * Optional custom configuration for the generated view of the page.
     */
    view?: {
        title?: string;
        path?: string;
    };

    /**
     * If set to true, then the generated route will point to a list of the entities, otherwise, it will point to a specific entity.
     */
    list?: boolean;
}

/**
 * Additional internal configuration options that will be calculated based on the provided input as well as the target application.
 */
export type InternalCustomPage = CustomPage &
    InternalCustomElement & {
        /**
         * The FLC flag is calculated to true if the existing app is configured to use the flexible column layout
         */
        fcl: boolean;

        /**
         * Represents the name of the control aggregation to be used for the custom page.
         * This is only used if FCL is enabled and is calculated based on the optional sourcepage.
         */
        controlAggregation?: 'beginColumnPages' | 'midColumnPages' | 'endColumnPages';
    };
