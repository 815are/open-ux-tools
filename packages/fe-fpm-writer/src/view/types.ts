import type { CustomElement, InternalCustomElement } from '../common/types';

export interface CustomView extends CustomElement {
    /**
     * Name of the routing target.
     */
    target: string;

    /**
     * View title, is shown in the tab item.
     */
    label: string;

    /**
     * Unique tab identifier.
     */
    key: string;

    /**
     * If not set (i.e. undefined) then no event handler is linked. If it is set true, a new one is created and linked to the action.
     * If an existing event handler is to be used then its id needs to be provided as string.
     */
    eventHandler?: string | true;

    /**
     * If set, a sample table control XML will be generated into the fragment of the view.
     */
    tableControl?: true;
}

export interface InternalCustomView extends CustomView, InternalCustomElement {
    content: string;
}
