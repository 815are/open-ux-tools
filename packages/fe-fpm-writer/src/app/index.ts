import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import { join } from 'path';
import { satisfies } from 'semver';
import type { SAPJSONSchemaForWebApplicationManifestFile as Manifest } from '../common/manifest';

export interface FPMConfig {
    fcl?: boolean;
}

/**
 * Enable the flexible programming model for an application.
 *
 * @param {string} basePath - the base path
 * @param {CustomTableColumn} config - the custom column configuration
 * @param {Editor} [fs] - the mem-fs editor instance
 * @returns {Promise<Editor>} the updated mem-fs editor instance
 */
export function enableFPM(basePath: string, config: FPMConfig = {}, fs?: Editor): Editor {
    if (!fs) {
        fs = create(createStorage());
    }

    const manifestPath = join(basePath, 'webapp/manifest.json');
    const manifest = fs.readJSON(manifestPath) as any as Manifest;

    // add FE libs is not yet added
    if (!manifest['sap.ui5']?.dependencies?.libs?.['sap.fe.templates']) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    libs: {
                        'sap.fe.templates': {}
                    }
                }
            }
        });
    }

    // if a minUI5Version is set and it is smaller than the minimum required, increase it
    if (
        manifest['sap.ui5']?.dependencies.minUI5Version &&
        !satisfies(manifest['sap.ui5']?.dependencies.minUI5Version, '>=1.86')
    ) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                dependencies: {
                    minUI5Version: '1.86.0'
                }
            }
        });
    }
    // enable FCL if requested
    if (config.fcl) {
        fs.extendJSON(manifestPath, {
            'sap.ui5': {
                rootView: {
                    viewName: 'sap.fe.templates.RootContainer.view.Fcl',
                    type: 'XML',
                    async: true,
                    id: 'appRootView'
                },
                routing: {
                    config: {
                        routerClass: 'sap.f.routing.Router'
                    }
                }
            }
        });
    }

    // replace Component.js
    const componentTemplate = join(__dirname, '../../templates/app/Component.js');
    fs.copyTpl(componentTemplate, join(basePath, 'webapp/Component.js'), manifest['sap.app']);

    return fs;
}
