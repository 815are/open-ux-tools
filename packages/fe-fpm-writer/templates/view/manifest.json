{
    "sap.ui5": {
        <% if (locals.eventHandler) { %> 
        "extends": {
            "extensions": {
                "sap.ui.controllerExtensions": {
                    "sap.fe.templates.ListReport.ListReportController": {
                        "controllerName": "<%- ns %>.<%- name %>"
                    }
                }
            }
        },
        <% } %>
        "routing": {
            "targets": {
                "<%- target %>": {
                    "options": {
                        "settings": {
                            "views": {
                                "paths": [
                                    {
                                        "key": "<%- key %>",
                                        "label": "<%- label %>",
                                        "template": "<%- ns %>.<%- name %>"
                                    }
                                ]
                            }
                        }
                    }
                }
            }
        }
    }
}