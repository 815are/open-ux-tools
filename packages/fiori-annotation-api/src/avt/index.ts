export { convertAnnotationFile } from './annotations';
export { convertMetadataToAvtSchema } from './metadata';
