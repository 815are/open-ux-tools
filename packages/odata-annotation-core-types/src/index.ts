export * from './annotation-file';
export * from './base';
export * from './edm';
export * from './edmx';

export { Location, Range, Position, Diagnostic } from './language-server';
