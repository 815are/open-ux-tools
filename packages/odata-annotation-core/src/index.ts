export * from '@sap-ux/odata-annotation-core-types';
export {
    ParsedName,
    ParsedActionFunctionSignature,
    ParsedCollectionIdentifier,
    ParsedIdentifier,
    parseIdentifier,
    toFullyQualifiedName
} from './names';

export {
    NavigationPropertyAnnotationSegment,
    ParsedPath,
    ParsedPathSegment,
    TermCastSegment,
    parsePath,
    toFullyQualifiedPath
} from './paths';
