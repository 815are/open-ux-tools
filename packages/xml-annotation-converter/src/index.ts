export { convertDocument, convertMetadataDocument } from './parser';
