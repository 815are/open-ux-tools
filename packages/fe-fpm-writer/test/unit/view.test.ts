import { create, Editor } from 'mem-fs-editor';
import { create as createStorage } from 'mem-fs';
import { join } from 'path';
import { generateCustomView } from '../../src';
import { CustomView } from '../../src/view/types';
import * as manifest from './sample/column/webapp/manifest.json';

const testDir = join(__dirname, 'sample/view');

describe('CustomView', () => {
    let fs: Editor;
    const customView: CustomView = {
        target: 'sample',
        key: 'viewKey',
        label: 'viewLabel',
        name: 'NewCustomView',
        folder: 'extensions/custom'
    };
    const expectedFragmentPath = join(testDir, 'webapp', customView.folder!, `${customView.name}.fragment.xml`);
    beforeEach(() => {
        fs = create(createStorage());
        fs.delete(testDir);
        fs.write(join(testDir, 'webapp/manifest.json'), JSON.stringify(manifest));
    });

    test('only mandatory properties', () => {
        //sut
        generateCustomView(testDir, customView, fs);
        const updatedManifest: any = fs.readJSON(join(testDir, 'webapp/manifest.json'));

        const views = updatedManifest['sap.ui5']['routing']['targets']['sample']['options']['settings']['views'];
        expect(views).toBeDefined();
        expect(views.paths).toBeDefined();
        expect(views.paths[0]).toMatchSnapshot();
        expect(fs.read(expectedFragmentPath)).toMatchSnapshot();
    });

    test('with new handler, all properties', () => {
        //sut
        generateCustomView(testDir, { ...customView, key: 'customViewKey', eventHandler: true }, fs);
        const updatedManifest: any = fs.readJSON(join(testDir, 'webapp/manifest.json'));

        const views = updatedManifest['sap.ui5']['routing']['targets']['sample']['options']['settings']['views'];
        expect(views).toBeDefined();
        expect(views.paths).toBeDefined();
        expect(views.paths[0]).toMatchSnapshot();
        expect(fs.read(expectedFragmentPath)).toMatchSnapshot();
        expect(fs.read(expectedFragmentPath.replace('.fragment.xml', '.js'))).toMatchSnapshot();
    });

    test('with custom control passed in interface', () => {
        const testCustomView: CustomView = {
            ...customView,
            control: '<CustomXML text="" />'
        };
        generateCustomView(testDir, testCustomView, fs);
        const updatedManifest: any = fs.readJSON(join(testDir, 'webapp/manifest.json'));

        const view =
            updatedManifest['sap.ui5']['routing']['targets']['sample']['options']['settings']['views']['paths'][0];
        expect(view).toMatchSnapshot();
        expect(fs.read(expectedFragmentPath)).toMatchSnapshot();
    });
});
