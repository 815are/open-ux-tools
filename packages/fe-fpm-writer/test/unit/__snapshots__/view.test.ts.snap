// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomView only mandatory properties 1`] = `
Object {
  "key": "viewKey",
  "label": "viewLabel",
  "template": "sapux.fe.fpm.writer.test.extensions.custom.NewCustomView",
}
`;

exports[`CustomView only mandatory properties 2`] = `
"<core:FragmentDefinition xmlns:core=\\"sap.ui.core\\" xmlns=\\"sap.m\\">
	<Text text=\\"NewCustomView\\" />
</core:FragmentDefinition>"
`;

exports[`CustomView with custom control passed in interface 1`] = `
Object {
  "key": "viewKey",
  "label": "viewLabel",
  "template": "sapux.fe.fpm.writer.test.extensions.custom.NewCustomView",
}
`;

exports[`CustomView with custom control passed in interface 2`] = `
"<core:FragmentDefinition xmlns:core=\\"sap.ui.core\\" xmlns=\\"sap.m\\">
	<CustomXML text=\\"\\" />
</core:FragmentDefinition>"
`;

exports[`CustomView with new handler, all properties 1`] = `
Object {
  "key": "customViewKey",
  "label": "viewLabel",
  "template": "sapux.fe.fpm.writer.test.extensions.custom.NewCustomView",
}
`;

exports[`CustomView with new handler, all properties 2`] = `
"<core:FragmentDefinition xmlns:core=\\"sap.ui.core\\" xmlns=\\"sap.m\\">
	<Button core:require=\\"{ handler: 'sapux/fe/fpm/writer/test/extensions/custom/NewCustomView'}\\" text=\\"NewCustomView\\" press=\\"handler.onPress\\" />
</core:FragmentDefinition>"
`;

exports[`CustomView with new handler, all properties 3`] = `
"sap.ui.define([
    \\"sap/m/MessageToast\\"
], function(MessageToast) {
    'use strict';

    return {
        onPress: function() {
            MessageToast.show(\\"Custom handler invoked.\\");
        }
    };
});
"
`;
